### ISSUE TITLE
Give user information on whether all CAR chunks got uploaded for larger uploads

### ISSUE THREAD

**Issue by dchoi27** [web3-storage/web3.storage#754]
Sister issue to: https://github.com/nftstorage/nft.storage/issues/896

In addition, #689 now shows everything in the uploads table as `Available`. This isn't actually true though - if there was a multipart upload for a file larger than 100MB, but not every chunk got uploaded (e.g. connection got interrupted), then the upload still shows up in the table even if not everything is retrievable in that situation. So we'll want to modify this as well.

**Comment by dchoi27** [comment]
Let's check the error message when not all CAR chunks get uploaded (not 100% sure if it's been implemented in both sides).

Otherwise brainstorming for this is here: https://hackmd.io/E0jxc0pCSxSkYgSUUq9lJA?view

**Comment by dchoi27** [comment]
Just a heads up - there's a website component to this that we'll probably only want to incorporate after website redesign launch. When we're creating tickets for this epic, let's label the front-end one with the topic/frontend label so we're reminded to tackle it after the redesign is launched. cc @orvn 

Also - when the CAR streaming work lands in NFT.Storage and we've ported it to Web3.Storage, figuring out what CARs are incomplete will get a lot easier for uploads moving forward

**Comment by orvn** [comment]
Is the goal here for the web UI to display how many chunks are uploaded and have a status `complete` visually, in the app UI? @dchoi27 

Btw, the `availability` column was added in the new UI, in parallel with https://github.com/web3-storage/web3.storage/pull/729

![availibility-web3s-files-table](https://user-images.githubusercontent.com/3689720/152891242-6b621de5-2703-4f87-a684-e3c6942c4997.png)

**Comment by dchoi27** [comment]
Cool! yeah - potentially more a warning if we don't think we have all the content (so doesn't have to be its own content, can just be something visually), but depends on the solution that potato decides on

**Comment by mikeal** [comment]
Why don’t we verify the upload is complete by asking for the last block in the last CAR file from a gateway after the upload?

**Comment by flea89** [comment]
**Hypothesis**
Read this in the context of the quick fix (to avoid the frustration of users thinking something is available while it is not) rather than the re-architecture of how uploads work.
I agree while that's required it definitively requires some more thoughts and changes to how control flow and data structures in the backend and defo has a dependency on the unix-fs-importer work.

> Why don’t we verify the upload is complete by asking for the last block in the last CAR file from an IPFS Gateway?

@mikeal [mikeal](https://github.com/mikeal) probably a silly question, why is this particularly different from relying on the status we get back from the cluster?

At the moment, we do already know an upload is not completed by the fact that `Pin Status !== pinned`, wouldn't be a quick win just to rely on that piece of information to display whether or not an upload is available?
It will be pessimistic ( I see there might be cases where the content is available but that status isn't showing as pinned but the content is already there), but that can be dealt with if needed.

**Comment by mikeal** [comment]
A few things:
1. pin status lags because we sync that state to the database asynchronously from ipfs-cluster.
2. the data may be available before being replicated to every node in the cluster, so it is “available” before the cluster status changes
3. i believe this column originally was just the pin status but it was causing confusion because of what i’ve mentioned above

i agree that it would be best to take a holistic approach to solving this problem. unfortunately, the holistic approach is to avoid pin status entirely, avoid using the root identifier in the CAR file entirely, and have the client maintain and update any state shared across multiple CAR files 😁

pin status is problematic for a number of reasons. as we move towards storage systems that rely purely on CAR files rather than pinning we’ll gain support for partial graphs. we’re in the process of that now, but in the meantime we’re using ipfs-cluster which is based around pinning. so it’s best if we can defer broader fixes until we have some other work in place, and we shouldn’t invest more than necessary in pin status.

**Comment by flea89** [comment]
Thanks for clarifying. 
I should have gone in a bit more depth in my comment, let me try to do it now while starting with your points.



Regarding
> Why don’t we verify the upload is complete by asking for the last block in the last CAR file from an IPFS Gateway?
> It would only take a couple of HTTP requests to get the root block and traverse down the right side until it hits a child node. If we can get that node then it’s a fair bet the file data is there since they are sent in order.

and 

> pin status lags because we sync that state to the database asynchronously from ipfs-cluster.

Thanks for suggesting this, it looks like a clever and “trustworthy” approach to get the availability. 
But wouldn’t this involve checking it asynchronously in a cron job, and/or checking it at every request to a “not-yet” available upload?
If that’s the case wouldn’t we have a similar problem to solve to the one you mention for pin statuses?
 
 
> the data may be available before being replicated to every node in the cluster, so it is “available” before the cluster status changes. the check i mentioned above would prove that reads to the end of the file are available from a gateway, which is a more honest availability check than pin status given gateways sometimes have issues finding or retrieving content
This is great context!

If I’m not mistaken, there’s a difference between a cluster giving a pinned status and a single node giving a pinned status.
The cluster APIs give us a granular, per peer status. If a single node tells us that the file is pinned then isn’t that a fairly reliable indication that the file is available? If that’s the case then would this be a good enough check of availability for now?
I think that the method of walking down the right-hand side of the tree is a great way to check the availability, but as an interim solution while we wait for the upload method to be refactored, would checking the pinned status be good enough to mitigate the problem that we’re seeing of users seeing their file as “Available” when it’s not?
 
The reason  I was suggesting this solution into consideration is that it requires minimal changes to control flow and none to data structure while fixing the user problem.
If we go down the gateway root, I think we need a bit more work and modify the current data structures to store the availability status after having it checked against the getaway. 
I wonder if that introduces more “concepts” that will have to get rid of/change anyway once we migrate to streaming and out of cluster pinning. 
That said, it’s not a lot of change, so we shouldn’t be crossing it off just for that.       

Another consideration that might make us lean towards one solution or the other:
- computational/time complexity differences between querying cluster status and querying/navigating dag through the getaway.


For context, at the moment, the dashboard is already displaying the [bestStatus](https://github.com/web3-storage/web3.storage/blob/main/packages/website/pages/files.js#L144) in the `Pinned Status` column, that means an upload would show as `Pinned` even if  the cluster reports only one node with `Pinned` status.
I wonder if a good approach would be to use `getBestPinStatus` to drive whether or not we show an upload as available. And consider having the PinStatus to say `Pinned` only when fully replicated?

Another approach which I think could probably avoid most of the confusion would be to keep the `Pin Status` column as is and get rid of the Available column altogether.
I’d love to hear other perspectives?

**Comment by olizilla** [comment]
"pin status" is us leaking internal state to the user that they can't meaningfully use. We need to give users a meaningful indication of _"is the entire dag from this root CID available from the gateway (http style) and via ipfs (p2p style)"_

**Comment by olizilla** [comment]
> Why don’t we verify the upload is complete by asking for the last block in the last CAR file

How would this work? The client is not stateful, so I think we'd have to post the CID of the "last" block to the API to keep track of it, which sounds like a bigger change than is wanted here, but might be a good direction to explore.

**Comment by flea89** [comment]
Thanks for looking at this @olizilla.

Can you clarify what you mean with:
> pinned: Show the count of nodes in the pinned status, or show the best pin status if it's not pinned on any.

Currently, the number of nodes is displayed in the tooltip, are you suggesting having the number be shown in the column cell instead when at least one node is reporting "pinned"? Or something different?

The big question here is: do we want the status "Pinned" to indicate full replication or not? 
If the answer is yes I assume we should wait for all the nodes to report a "Pinned" status.
If not it's probably ok to display "Pinned"  in the cell and the number of nodes in the tooltip?

I think in either case we should probably update the header tooltip to make that obvious to the user.

**Comment by olizilla** [comment]
I'd need to implement it and see how it felt to know what would be best for the pinned column. We can just fix the availble column for now and leave pinned as it is. I think the specific change being discussed now is small, so let's get it fixed and we can move on to designing something better.

**Comment by dchoi27** [comment]
So some background: We used to have just the `Pin Status` column, but `Available` was added later because pin statuses were taking a long time to update, and a TON of users were asking about this (they weren't sure if their data was safe, or if the gateway was having infra issues and the pin status still said queuing they thought it was a problem with their upload rather than the gateway infra being the problem).

I actually didn't know that we were using `bestStatus` to report pin status now, but at the time we introduced "Available," user uploads were stuck in `Queuing` for ages. Maybe using `bestStatus` was introduced later - I do remember the convos at the time centering around Cluster taking a long time to propagate content to all nodes, so `bestStatus` might have come after that. Or maybe we missed that `bestStatus` was being used and it wasn't a Cluster problem (maybe the cronjob taking a long time to update things, or the individual nodes were also behind in updating pin status). 

It does sound like #1009 is just a simplified way of doing what we were doing before where we have fewer false positives (since "Available" is gone, and we no longer say that incomplete uploads are available) but potentially more false negatives (if things don't go from "Queued" to "Pinned" quickly enough). Unless we know that there have been improvements shifting things from Queued to Pinned since the "Available" column was introduced (e.g., bestStatus being used, cronjob updating things faster/more consistently, etc.), I think we just end up exposing ourselves to what we did before (users see everything says "Queued" and this is the only info they have to go off of) - #912 for instance shows that it might not be good to just rely on the pin status updating.

In any case, this ticket was opened to try and minimize the false positives. This applies whether or not we have the "Available" column - things stuck in Queuing is also a false positive. We wanted to explore ways to figure out whether we actually have all pieces of an upload while we're waiting for the full streaming CAR work to finish (ideally the solution here would work for both current and future CAR generation flows). So let's make sure not to close this ticket until @alanshaw's or @mikeal's solutions are implemented, or if we decide to just wait until the CAR streaming work is done.

One flag - I agree with all the conversation about abstracting our infra's status information as much as we can from the user, but I think #1009 probably makes us rely on it more since, once again, the only info that will be displayed is what our infra is saying. Again, that's probably OK if we think `Queuing` will be updated more quickly / more often to `Pinned` vs. what we saw before "Availability" was introduced.

**Comment by dchoi27** [comment]
Spoke with @alanshaw - he wasn't aware of the `Availability` column and we agreed we should take this down ASAP (i.e., the false positives are much worse than false negatives). So #1009 (or #1026, whatever gets merged) shouldn't close this issue (#754). We could also consider bringing back `Availability` once this issue (#754) is closed.

cc @orvn - this affects how we're displaying things in the front-end at launch!

**Comment by flea89** [comment]
We shouldn't have closed this issues, according to David's [suggestion](https://github.com/web3-storage/web3.storage/issues/754#issuecomment-1048051048).

Opening again.

**Comment by fantompocong** [comment]
If the status is **Queuing** can I turn off my laptop? or it should stay on until **Pinned**?? Been refreshing all day and for a short period of time I saw the Availability column said it Pinned but the status still Queuing, then i refresh it again the column is gone. Please delete if this comment irrelevant. Thanks in advance.

**Comment by dchoi27** [comment]
How did you upload the file? There should be no more Availability column (and it never should have said `Pinned`, just `Available`. But if you uploaded directly to Web3.Storage (instead of pinning from your local node for instance) then it should be good (one way to check is if you can access the data via an IPFS gateway).

**Comment by fantompocong** [comment]
> 

Following this https://docs.web3.storage/#create-the-upload-script as is. But i uploaded a folder so instead of `node put-files.js --token=<MY_TOKEN> ~/filename1 ~/filename2` i used `node put-files.js --token=<MY_TOKEN> my-directory-name`.

i can see the folder using my-cid.ipfs.dweb.link but when i add my-cid.dweb.link/1.png it said "no link named "1.png" under my-cid.

**Comment by dchoi27** [comment]
Oh - that sounds like the upload didn't complete (so state will never exit from Queuing). I would try reuploading again, and once you've confirmed it's completed, the entire directory should be available via the gateway.

Know that Queuing is ambiguous (is it because the upload didn't complete? or because it hasn't finished redundantly pinning?) - so this issue #754 is trying to give the user more specific information around that.

**Comment by fantompocong** [comment]
> Oh - that sounds like the upload didn't complete (so state will never exit from Queuing). I would try reuploading again, and once you've confirmed it's completed, the entire directory should be available via the gateway.
> 
> Know that Queuing is ambiguous (is it because the upload didn't complete? or because it hasn't finished redundantly pinning?) - so this issue #754 is trying to give the user more specific information around that.

probably? on terminal inside VSC it still saying "Uploading 7777 files", been running more than 12 hours. Oh, when I said I saw "Pinned", I also saw the Size which 2.9 GB, i refresh it then it back to Calculating....

Also... before using this method (to upload to web3.storage), on other terminal window i tried to upload using 
`nftp` from `npm install -g nftp` to nft.storage, until now it still saying "uploading" but all the file already can be access, the only problem is it failed to pin.

both uploaded the same folder.

**Comment by dchoi27** [comment]
Oh hm, 12 hours is crazy. What are you using to upload - sounds like potentially a script?

`nftp` is a community project so I can't personally speak to that!

Mind opening a new issue with some more details on what you're doing, what kinds of files / structure you're uploading, etc.? Since this issue is separate.

**Comment by fantompocong** [comment]
> Oh hm, 12 hours is crazy. What are you using to upload - sounds like potentially a script?
> 
> `nftp` is a community project so I can't personally speak to that!
> 
> Mind opening a new issue with some more details on what you're doing, what kinds of files / structure you're uploading, etc.? Since this issue is separate.

Yeah, using the script i mentioned above. I actually talk on behalf of my friend, will try to recreate it on my own account. If the problem still there I'll create a new issue. Thanks for answering. Also if our convo is irrelevant we can delete it from here.
